submitter/tx_submitter.go

```diff
@@ -20,151 +22,177 @@ import (
 func (s *TxSubmitter) SubmitTransactionLoop() {
-	// Event lasts for 300 blocks, 2x for redundancy
-	s.cachedEventHash = make(map[uint64][]byte, common.CacheSize)
 	submitLoopCount := 0
-	for {
-		attestPeriodEnd := uint64(0)
-		for {
-			// blsPubKey of current submitter
-			res, err := s.executor.QueryInturnAttestationSubmitter()
-			if err != nil {
-				logging.Logger.Errorf("tx submitter failed to query inturn attestation submitter, err=%+v", err.Error())
-				continue
-			}
-			attestPeriodEnd = res.SubmitInterval.GetEnd()
-			if res.BlsPubKey == hex.EncodeToString(s.executor.BlsPubKey) {
-				logging.Logger.Infof("tx submitter is currently inturn for submitting until: %s, current timestamp: %s", time.Unix(int64(attestPeriodEnd), 0).Format("15:04:05.000000"), time.Now().Format("15:04:05.000000"))
-				break
-			}
-			time.Sleep(common.RetryInterval)
-			continue
-		}
 
+	for {
+		// Loop until submitter is inturn to submit
+		attestPeriodEnd := s.queryAttestPeriodLoop()
+		// Fetch events for submit
 		currentHeight := s.executor.GetCachedBlockHeight()
 		events, err := s.FetchEventsForSubmit(currentHeight)
-		logging.Logger.Infof("tx submitter fetched %d events for submit", len(events))
 		if err != nil {
-			logging.Logger.Errorf("tx submitter failed to fetch events for submitting, err=%+v", err.Error())
+			logging.Logger.Errorf("tx submitter failed to fetch events for submitting", err)
 			continue
 		}
 		if len(events) == 0 {
 			time.Sleep(common.RetryInterval)
 			continue
 		}
-
+		// Submit events
 		for _, event := range events {
 			err = s.submitForSingleEvent(event, attestPeriodEnd)
 			if err != nil {
-				logging.Logger.Errorf("tx submitter err=%+v, timestamp: %s", err.Error(), time.Now().Format("15:04:05.000000"))
+				logging.Logger.Errorf("tx submitter err", err)
 				continue
 			}
-			time.Sleep(50 * time.Millisecond)
+			time.Sleep(TxSubmitInterval)
 		}
-		// clear event hash every N attempts
+		// Clear cached event hash
 		submitLoopCount++
 		if submitLoopCount == common.CacheClearIterations {
 			submitLoopCount = 0
-			s.cachedEventHash = make(map[uint64][]byte, common.CacheSize)
+			s.cachedEventHash = make(map[uint64][]byte, 0)
+		}
+
+		time.Sleep(TxSubmitLoopInterval)
+	}
+}
+
+// queryAttestPeriodLoop loops until submitter is inturn and return the end time of the current attestation period.
+func (s *TxSubmitter) queryAttestPeriodLoop() uint64 {
+	for {
+		res, err := s.executor.QueryInturnAttestationSubmitter()
+		if err != nil {
+			continue
+		}
+		// Submitter is inturn if bls key matches
+		if res.BlsPubKey == hex.EncodeToString(s.executor.BlsPubKey) {
+			logging.Logger.Infof("tx submitter is currently inturn for submitting until", time.Unix(int64(res.SubmitInterval.GetEnd()), 0).Format(TimeFormat))
+			return res.SubmitInterval.GetEnd()
 		}
-		time.Sleep(executor.QueryAttestedChallengeInterval)
+
+		time.Sleep(common.RetryInterval)
```

Comment:
actually, we can use the end time in `res` to determine how long we should wait.